hack/generator/pkg/astbuilder/builder.go

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astbuilder
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+)
+
+// CheckErrorAndReturn checks if the err is non-nil, and if it is returns.
+func CheckErrorAndReturn(otherReturns ...*ast.Ident) ast.Stmt {
+
+	var returnValues []ast.Expr
+	for _, ret := range otherReturns {
+		returnValues = append(returnValues, ret)
+	}
```

Comment:
`otherReturns[:]` won't create new storage, just a new slice object (https://play.golang.org/p/zmyDBLe9vGu) - you'd need to use `copy` to get that:

https://play.golang.org/p/NWN7LRtiLIQ

Another option is `returnValues := append([]ast.Expr, otherReturns...)`.

Reply:
Huh. I tried it out in the Playground before making the comment, so _sometimes_ the storage isn't shared. But I can't now reproduce it. 😁

Reply:
Well, it creates a new slice object (pointer, length, capacity), but the pointer will be shared. It might not look like they're shared if you do an append when length = capacity, because in that case it needs to allocate new storage and copy the existing items across and it will only update the pointer the new slice b.
https://play.golang.org/p/jorKrpM7vb2

But if length < capacity then the append won't reallocate, so a and b will both be operating on the same storage.

https://play.golang.org/p/Xcb5YvQMxE8

Reply:
I think the original issue doing what @babbageclunk had suggested was that I had the parameter type here as `[]*ast.Ident` but the collection to supply to `ast.ReturnStmt` needed to be an `[]ast.Expr`, and you can't "reinterpret" a collection type like that in Go (the generics they have are not covariant).

In retrospect I should've just made the parameter `[]ast.Expr` instead as that's more flexible anyway (and still works for `*ast.Ident` since `*ast.Ident` impls `ast.Expr`.

Reply:
Oh right, I hadn't noticed this was a type conversion as well.